test/unittests/execute_floating_point_test.cpp

```diff
@@ -42,6 +42,134 @@ MATCHER_P(ArithmeticNaN, value, "result with an arithmetic NaN")
+        T{0.0},
+
+        Limits::denorm_min(),
+        std::nextafter(Limits::denorm_min(), Limits::infinity()),
+        std::nextafter(Limits::min(), T{0}),
+        Limits::min(),
+        std::nextafter(Limits::min(), Limits::infinity()),
+        std::nextafter(T{1.0}, T{0}),
+        T{1.0},
+        std::nextafter(T{1.0}, Limits::infinity()),
+        std::nextafter(Limits::max(), T{0}),
+        Limits::max(),
+
+        Limits::infinity(),
+
+        // Canonical NaN:
+        FP<T>::nan(FP<T>::canon),
+
+        // Arithmetic NaNs:
+        FP<T>::nan((FP<T>::canon << 1) - 1),             // All bits set.
+        FP<T>::nan(FP<T>::canon | (FP<T>::canon >> 1)),  // Two top bits set.
+        FP<T>::nan(FP<T>::canon + 1),
+
+        // Signaling (not arithmetic) NaNs:
+        FP<T>::nan(FP<T>::canon >> 1),  // "Standard" signaling NaN.
+        FP<T>::nan(2),
+        FP<T>::nan(1),
+    };
+
+public:
+    using Iterator = typename decltype(m_values)::const_iterator;
+
+    static constexpr auto num_nans = 7;
+    static constexpr auto num_positive = m_values.size() - num_nans;
+    static constexpr auto num_ordered = 2 * num_positive - 1;
+
+    class Range
+    {
+        Iterator m_begin;
+        Iterator m_end;
+
+    public:
+        constexpr Range(Iterator begin, Iterator end) noexcept : m_begin{begin}, m_end{end} {}
+
+        constexpr Iterator begin() const { return m_begin; }
+        constexpr Iterator end() const { return m_end; }
+        constexpr size_t size() const { return static_cast<size_t>(m_end - m_begin); }
+    };
+
+    // The list of positive floating-point values without zero, infinity and NaNs.
+    static Range positive_nonzero_finite() noexcept
+    {
+        return {&m_values[1], &m_values[num_positive - 1]};
+    }
+
+    // The list of positive floating-point values without zero and NaNs (includes infinity).
+    static Range positive_nonzero_infinite() noexcept
+    {
+        return {&m_values[1], &m_values[num_positive]};
+    }
+
+    // The list of positive NaN values.
+    static Range positive_nans() noexcept { return {m_values.end() - num_nans, m_values.end()}; }
+
+    // The list of positive non-canonical NaN values (including signaling NaNs).
+    static Range positive_noncanonical_nans() noexcept
+    {
+        return {m_values.end() - num_nans + 1, m_values.end()};
+    }
+
+    // The list of positive floating-point values with zero, infinity and NaNs.
+    static Range positive_all() noexcept { return {m_values.begin(), m_values.end()}; }
+
+    // The list of floating-point values, including infinities.
+    // They are strictly ordered (ordered_values[i] < ordered_values[j] for i<j).
+    // Therefore -0 is omitted. This allows determining the relation of any pair of values only
+    // knowing values' position in the array.
+    static auto& ordered() noexcept
+    {
+        static const auto ordered_values = [] {
+            std::array<T, num_ordered> a;
```

Comment:
I would instead of declaring `num_ordered` above, calculate it here as `ps.size() * 2  + 1`